cody-repo-util.el

```diff
@@ -0,0 +1,182 @@
+                                 (cody--get-root-file-from-first-git-repo project))))
+    (condition-case nil
+        (cody--get-remote-repo-url project file-from-repo)
+      (error
+       (cody--get-simple-repository-name project file-from-repo)))))
+
+(defun cody--get-simple-repository-name (project file)
+  "Get the simple repository name for PROJECT and FILE."
+  (when-let ((repository (vc-call-backend (vc-backend file) 'root file)))
+    (file-name-nondirectory (directory-file-name repository))))
+
+(defun cody--url-do-replacements (remote-url)
+  "Perform URL replacements for REMOTE-URL"
+  (let* ((replacements
+          (split-string cody--remote-url-replacements "\\s*,\\s*"))
+         (remote-url-with-replacements remote-url))
+    (when (zerop (% (length replacements) 2))
+      (cl-loop for i from 0 to (1- (/ (length replacements) 2))
+               do (setq remote-url-with-replacements
+                        (replace-regexp-in-string
+                         (nth (* 2 i) replacements)
+                         (nth (1+ (* 2 i)) replacements)
+                         remote-url-with-replacements))))
+    remote-url-with-replacements))
+
+(defun cody--get-remote-repo-url (project file)
+  "Get the remote repository URL for PROJECT and FILE.
+Returned format: github.com:sourcegraph/sourcegraph.git"
+  (let ((repository (vc-call-backend (vc-backend file) 'root file))
+        (vcs-type (cody--get-vcs-type project file)))
+    (cond
+     ((and (eq vcs-type 'git) repository)
+      (cody--convert-git-clone-url-to-codebase-name (vc-git-repository-url repository)))
+     ((eq vcs-type 'perforce)
+      (cody--get-perforce-remote-repo-url project file))
+     (t (error "Unsupported VCS or repository not found for file %s" file)))))
+
+(defun cody--get-repo-root-path (project file)
+  "Get the repository root directory for PROJECT and FILE."
+  (let ((vcs-root (vc-call-backend (vc-backend file) 'root file)))
+    (when vcs-root
+      (file-name-directory vcs-root))))
+
+(defun cody--get-remote-branch-name (project file)
+  "Get the remote branch name for PROJECT and FILE."
+  (let ((repository (vc-call-backend (vc-backend file) 'root file)))
+    (when (and repository (eq (cody--get-vcs-type project file) 'git))
+      (vc-git-working-revision repository))))
+
+(defun cody--get-vcs-type (project file)
+  "Get the VCS type for PROJECT and FILE."
+  (let ((repository (vc-call-backend (vc-backend file) 'root file)))
+    (cond
+     ((and (require 'vc-git nil t)
+           (vc-call-backend 'Git 'root file)) 'git)
+     ((and (require 'p4 nil t)
+           (p4-get-client-name)) 'perforce)
+     (t 'unknown))))
+
+(defun cody--get-root-file-from-first-git-repo (project)
+  "Get the root file from the first Git repository in PROJECT."
+  (let ((repo (locate-dominating-file project ".git")))
+    (when repo
+      (expand-file-name repo))))
+
+(defun cody--convert-git-clone-url-to-codebase-name (url)
```

Comment:
For the TypeScript extension I just ported something from Go which did this and it seemed to be long-lived. Though NOW I'm told that there's a GraphQL for converting clone URLs to codebase names and admins can create a mapping, so the truth might depend whether you want to pass one of these names to a remote (and most of the time what you have here will be right anyway...)

Reply:
yeah most of the time this will be right. On sourcegraph server side we do have a way which will convert things in the sourcegraph name based on remote. Indeed admins can set up templates for how to generate names. eg some instances will leave off the hostname from repo name to make nicer urls.

Reply:
I'd be happy to make a graphql call to get this correctly - will take a look.

But maybe after we merge!